src/core/namespace.c

```diff
@@ -1246,22 +1266,32 @@ static int make_read_only(const MountEntry *m, char **deny_list, FILE *proc_self
                 flags_mask |= MS_NOSUID;
         }
 
+        if (mount_entry_noexec(m)) {
+                new_flags |= MS_NOEXEC;
+                flags_mask |= MS_NOEXEC;
+                submounts = true;
+        } else if (mount_entry_exec(m)) {
+                new_flags &= ~MS_NOEXEC;
+                flags_mask |= MS_NOEXEC;
+                submounts = true;
+        }
+
         if (flags_mask == 0) /* No Change? */
                 return 0;
 
         /* We generally apply these changes recursively, except for /dev, and the cases we know there's
          * nothing further down.  Set /dev readonly, but not submounts like /dev/shm. Also, we only set the
          * per-mount read-only flag.  We can't set it on the superblock, if we are inside a user namespace
          * and running Linux <= 4.17. */
-        submounts =
+        submounts |=
                 mount_entry_read_only(m) &&
                 !IN_SET(m->mode, EMPTY_DIR, TMPFS);
         if (submounts)
                 r = bind_remount_recursive_with_mountinfo(mount_entry_path(m), new_flags, flags_mask, deny_list, proc_self_mountinfo);
         else
                 r = bind_remount_one_with_mountinfo(mount_entry_path(m), new_flags, flags_mask, proc_self_mountinfo);
```

Comment:
so, this doesn't look right: this way if noexec or exec is set for a dir, the read-only flag that was supposed to be applied only to the dir itself suddenly gets applied recursively too...

Also the EMPTY_DIR/TMPFS check kinda applies to noexec/exec too

Reply:
fixed